contrib/cgi/src/cgi.c

```diff
@@ -0,0 +1,723 @@
+	if (bytes_read == -1 || bytes_read != read_len) {
+		if (!error && !*error)
+			g_set_error(error, CGI_ERROR, CGI_ERROR_BAD_REQUEST, "Content-Length header incorrect.");
+		write_upload_status(-1, NULL);
+		goto error;
+	}
+
+	if (!g_file_set_contents(BUNDLE_TARGET_LOCATION, file_buf, read_len, error)) {
+		write_upload_status(-1, NULL);
+		goto error;
+	}
+
+	ret = TRUE;
+
+error:
+	if (*error)
+		/* try to remove status/lock file */
+		g_remove(STATUS_FILE_LOCATION);
+
+out:
+	g_free(file_buf);
+	g_free(content_length);
+
+	return ret;
+}
+
+/*
+ * Handle CGI env variables and trigger requested operations.
+ */
+static gint cgi_handler(int argc, char **argv)
+{
+	gint ret = 1;
+	RInstaller *installer;
+	gchar *method = g_strdup(g_getenv("REQUEST_METHOD"));
+	gchar *query_string = g_strdup(g_getenv("QUERY_STRING"));
+	GError *error = NULL;
+
+	installer = r_installer_proxy_new_for_bus_sync(G_BUS_TYPE_SYSTEM, G_DBUS_PROXY_FLAGS_GET_INVALIDATED_PROPERTIES, "de.pengutronix.rauc", "/", NULL, &error);
+	if (!installer)
+		goto error;
+
+	/*
+	 * there is no way to remove stale bundle/status files directly after
+	 * install failure, so just do it now
+	 */
+	try_cleanup_status_on_install_failure(installer);
+
+	if (g_strcmp0("GET", method) == 0) {
+		if (g_strcmp0("progress", query_string) == 0) {
+			/* ?progress shows progress information */
+			if (!progress_status(installer, &error))
+				goto error;
+
+		} else if (g_strcmp0("status", query_string) == 0) {
+			/* ?status shows slot information */
+			if (!slot_status(installer, &error))
+				goto error;
+		} else {
+			g_set_error(&error, CGI_ERROR, CGI_ERROR_BAD_REQUEST, "Resource '%s' unavailable", query_string);
+			goto error;
+		}
+
+	} else if (g_strcmp0("PUT", method) == 0) {
+		/* save stdin to temporary file */
+		if (!stdin_to_file(&error))
+			goto error;
+
+		/* start rauc install */
+		if (!r_installer_call_install_sync(installer, BUNDLE_TARGET_LOCATION, NULL, &error))
+			goto error;
+
+		print_headers("200 OK", "text/plain");
+		g_print("Upload and install trigger executed successfully.\n");
+
+	} else {
+		g_set_error(&error, CGI_ERROR, CGI_ERROR_METHOD_NOT_ALLOWED, "Unsupported method '%s'", method);
+		goto error;
+	}
+
+	ret = 0;
+error:
+	if (error) {
+		/* try to remove bundle */
+		g_remove(BUNDLE_TARGET_LOCATION);
+
+		switch (error->code) {
+			case CGI_ERROR_METHOD_NOT_ALLOWED:
+				print_headers("405 Method Not Allowed", "text/plain");
+				break;
+			case CGI_ERROR_BAD_REQUEST:
+				print_headers("400 Bad Request", "text/plain");
+				break;
+			case CGI_ERROR_LENGTH_REQUIRED:
+				print_headers("411 Length Required", "text/plain");
+				break;
+			case CGI_ERROR_SERVICE_UNAVAILABLE:
+				print_headers("503 Service Unavailable", "text/plain");
+				break;
+			default:
+				print_headers("500 Internal Server Error", "text/plain");
+		}
+		g_printerr("%s\n", error->message);
+		g_error_free(error);
+	}
+
+out:
```

Comment:
Not used and can be removed.

Reply:
Yes, I left it there for structural reasons. Will remove it.